flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/codegen/agg/ImperativeAggCodeGen.scala

```diff
@@ -0,0 +1,502 @@
+        var term = s"${genToExternal(ctx, externalUDITypes(index), inputExpr.resultTerm)}"
+        s"${inputExpr.nullTerm} ? null : $term"
+      }
+    }
+
+    val accTerm = if (isAccTypeInternal) accInternalTerm else accExternalTerm
+    // insert acc to the head of the list
+    val fields = Seq(accTerm) ++ inputFields
+    // acc, arg1, arg2
+    (fields.mkString(", "), codes.mkString("\n"))
+  }
+
+  /**
+    * This method is mainly the same as CodeGenUtils.generateFieldAccess(), the only difference is
+    * that this method using UpdatableRow to wrap BaseRow to handle DataViews.
+    */
+  def generateAccumulatorAccess(
+    ctx: CodeGeneratorContext,
+    inputType: InternalType,
+    inputTerm: String,
+    index: Int,
+    viewSpecs: Array[DataViewSpec],
+    useStateDataView: Boolean,
+    useBackupDataView: Boolean): GeneratedExpression = {
+
+    // if input has been used before, we can reuse the code that
+    // has already been generated
+    val inputExpr = ctx.getReusableInputUnboxingExprs(inputTerm, index) match {
+      // input access and unboxing has already been generated
+      case Some(expr) => expr
+
+      // generate input access and unboxing if necessary
+      case None =>
+        // this field access is not need to reuse
+        val expr = generateFieldAccess(ctx, inputType, inputTerm, index)
+
+        val newExpr = inputType match {
+          case ct: RowType if isAccTypeInternal =>
+            // acc is never be null
+            val fieldType = ct.getTypeAt(index).asInstanceOf[RowType]
+            val exprGenerator = new ExprCodeGenerator(ctx, false)
+              .bindInput(fieldType, inputTerm = expr.resultTerm)
+            val converted = exprGenerator.generateConverterResultExpression(
+              fieldType,
+              classOf[GenericRow],
+              outRecordTerm = newName("acc"),
+              reusedOutRow = false,
+              fieldCopy = inputFieldCopy)
+            val code =
+              s"""
+                 |${expr.code}
+                 |${ctx.reuseInputUnboxingCode(expr.resultTerm)}
+                 |${converted.code}
+               """.stripMargin
+
+            GeneratedExpression(
+              converted.resultTerm,
+              converted.nullTerm,
+              code,
+              converted.resultType)
+          case _ => expr
+        }
+
+        val exprWithDataView = inputType match {
+          case ct: RowType if viewSpecs.nonEmpty && useStateDataView =>
+            if (isAccTypeInternal) {
+              val code =
+                s"""
+                  |${newExpr.code}
+                  |${generateDataViewFieldSetter(newExpr.resultTerm, viewSpecs, useBackupDataView)}
+                """.stripMargin
+              GeneratedExpression(newExpr.resultTerm, newExpr.nullTerm, code, newExpr.resultType)
+            } else {
+              val fieldType = ct.getTypeAt(index)
+              val fieldTerm = ctx.newReusableField("field", UPDATABLE_ROW)
```

Comment:
I would suggest to use `val fieldTerm  = ctx.addReusableLocalVariable(UPDATABLE_ROW, "field")` instead to avoid introduce a new method `newReusableField`.

The local variables will be add to member area if code need to be split, otherwise will be put in method as local variable.